programs/utilities/keepalive.sc

```diff
@@ -1,21 +1,40 @@
 // re-logs fake players upon server restart
 // must need to place it directly in the world scripts folder
 
-__config() -> {'scope' -> 'global'};
+global_cached_players = {};
+
+__config() -> {
+	requires -> {
+		'minecraft' -> '>=1.8', // spectator was not a thing prior to 1.8
+	},
+	'scope' -> 'global'
+};
+
+__on_connect(e) -> (
+	// forcing the fake-players that were spawned in spectator back into creative to retain flight per their data
+	if(e:'gm' == 'creative' && e:'fly' == 1,
+	// scheduling the modification at the end to ensure the entity was created properly and resetting it back to the intended gamemode
+		schedule(0, _(e) -> (modify(player(e:'name'), 'flying', e:'fly')), e);
+	);
+);
+
+__on_player_connects(player) -> (
+	player_name = player()~'name';
+
+	if (has(global_cached_players, player_name),
+		task('__on_connect', global_cached_players:player_name)
```

Comment:
Don't see a reason to use task - theres no expensive conputation or sleep inside. It schedules immediately anyway. In that case you could also just move the contents of `__on_connect` into the `player_connects` event directly.
In fact if using the `modify(e, 'flying'` approach (which i prefer) using a task here could lead to some delay before setting the flying, during which the player could be ticked, causing the player to drop down a bit.

I would consider removing the `task` for `__spawn_players` in `__on_server_starts` too. This seems to have been a fix for this script previously (https://github.com/gnembon/scarpet/pull/410) seemingly not working for everyone. Since this should now be fixed fully by waiting for the player to be spawned with the event, it should be fine to remove.

Reply:
In my attempts, replacing `task('__on_connect', global_cached_players:player_name);` with `__on_connect(global_cached_players:player_name);` worked fine. As long as there is still the `schedule(0,...)` in there, it will run the `modify` on the main thread (end of tick, after the player fully connected) anyway.